fastpath for `a:=mat[i,j]` and `mat[i,j]:=a`

fingolfin opened the pull request:
This PR is work in progress, and was started during the MatrixObj GAP Days in Aachen.

This does several things:
* adds code to allow writing `mat[i,j]` as an alias for `mat[i][j]`
* adds new kernel methods to accessing entries of compressed matrices
* adds a set of benchmarks for the various ways to access matrix entries (this benchmark will be either removed in the final version; or perhaps be kept, but then renamed and moved to a more appropriate location; for now I keep it as `bench-mat.g` in the root directory of the repos, for my convenience). The benchmark tests various ways to access entries of various kinds of matrices:
  1. `mat[i][j]`
  2. `mat[i,j]`
  3. `MatElm(mat,i,j)`
  4. `MatElm(mat,i,j)` with prefetched method (to avoid method selection)

With this PR, using `mat[i,j]` actually is always as fast or even faster than `mat[i][j]`, both for reading and writing matrix entries; for plists of plists, for plists of compressed vectors, for compressed matrices (except prefetched `SetMatElm` for a GF(7) matrix -- but (a) you can also prefetch the `[]:=` method, and (b) I can add 8bit matrices to the hotpath, just like GF(2) matrices, and it'll be faster again). Here are some benchmarking results (using GCC 7, on my 2012 MacBook Pro):
```
+------------------------+
| Testing integer matrix |
+------------------------+
Testing m[i][j]:
  260 iterations per second
Testing m[i,j]:
  287 iterations per second
Testing MatElm(m,i,j):
  18 iterations per second
Testing MatElm(m,i,j) with prefetched method:
  91 iterations per second
...now testing write access...
Testing m[i][j]:=elm:
  231 iterations per second
Testing m[i,j]:=elm:
  298 iterations per second
Testing SetMatElm(m,i,j,elm):
  18 iterations per second
Testing SetMatElm(m,i,j,elm) with prefetched method:
  81 iterations per second

+-----------------------+
| Testing GF(2) rowlist |
+-----------------------+
Testing m[i][j]:
  184 iterations per second
Testing m[i,j]:
  222 iterations per second
Testing MatElm(m,i,j):
  24 iterations per second
Testing MatElm(m,i,j) with prefetched method:
  78 iterations per second
...now testing write access...
Testing m[i][j]:=elm:
  149 iterations per second
Testing m[i,j]:=elm:
  158 iterations per second
Testing SetMatElm(m,i,j,elm):
  23 iterations per second
Testing SetMatElm(m,i,j,elm) with prefetched method:
  71 iterations per second

+---------------------------------+
| Testing GF(2) compressed matrix |
+---------------------------------+
Testing m[i][j]:
  147 iterations per second
Testing m[i,j]:
  195 iterations per second
Testing MatElm(m,i,j):
  177 iterations per second
Testing MatElm(m,i,j) with prefetched method:
  194 iterations per second
...now testing write access...
Testing m[i][j]:=elm:
  108 iterations per second
Testing m[i,j]:=elm:
  162 iterations per second
Testing SetMatElm(m,i,j,elm):
  139 iterations per second
Testing SetMatElm(m,i,j,elm) with prefetched method:
  181 iterations per second

+-----------------------+
| Testing GF(7) rowlist |
+-----------------------+
Testing m[i][j]:
  150 iterations per second
Testing m[i,j]:
  142 iterations per second
Testing MatElm(m,i,j):
  24 iterations per second
Testing MatElm(m,i,j) with prefetched method:
  74 iterations per second
...now testing write access...
Testing m[i][j]:=elm:
  91 iterations per second
Testing m[i,j]:=elm:
  93 iterations per second
Testing SetMatElm(m,i,j,elm):
  22 iterations per second
Testing SetMatElm(m,i,j,elm) with prefetched method:
  53 iterations per second

+---------------------------------+
| Testing GF(7) compressed matrix |
+---------------------------------+
Testing m[i][j]:
  118 iterations per second
Testing m[i,j]:
  153 iterations per second
Testing MatElm(m,i,j):
  141 iterations per second
Testing MatElm(m,i,j) with prefetched method:
  148 iterations per second
...now testing write access...
Testing m[i][j]:=elm:
  73 iterations per second
Testing m[i,j]:=elm:
  78 iterations per second
Testing SetMatElm(m,i,j,elm):
  78 iterations per second
Testing SetMatElm(m,i,j,elm) with prefetched method:
  90 iterations per second
```

Things that need to be done:
* [ ] add comprehensive tests for everything
* [ ] clean up various hacks
* [ ] document the `[i,j]` syntax in the manual
* [ ] in light of the benchmark results, talk with folks (on mailing list?) about MatrixObj; in particular, whether we should remove `MatElm` and `SetMatElm` from the `MatrixObj` API again, instead requiring that suitable `[]` and `[]:=` methods be provided.
* [ ] Discuss what to do with `foo[a,b,c]` with more than two arguments. Previously, we allowed that, by stuffing all indices into a plist. With this PR, I instead invoke the `[]` and `[]:=` operations with additional arguments for each index (and in fact, only 1 or 2 indices are currently supported in this PR). The reasons for this change are: 1. we avoid creating a temporary plist, 2. it becomes easier to install methods with precise filters, 3. we avoid having `IsList` filters in the method selection for `[]` and `[]:=` -- and avoiding the overhead for `IsList` in method selection is one of the driving forces of the MatrixObj project. My inclination right now is to have 3 or more indices trigger an error; and if anybody ever wants to use 3 or more indices for some project, we can re-activate it then, in a manner that makes sense for that actual use case.
* [ ] Decide on semantics for out-of-bounds access; then implement it, add tests, ensure error messages are consistent and helpful.
* [ ] react to feedback y'all have on this work :-)

stevelinton commented:
The fastpath stuff looks good. It is what I had in mind when I split ASS2 and ELM2 out in the parser. 

The question of what to do with three or more indices is an interesting one. There are certainly places in the code where one sees X[i][j][k]...[z] with three or more indices, albeit not very many of them. If we go down the route of installing methods with more and more arguments we are limited to 4 indexes for assignment because of the limit to six arguments for operations. Maybe that's enough.

fingolfin commented:
Of course we could also decide to treat the case of more than N indices (where N could be 2, or 4, or whatever) differently, and for those, pass the indices in the list... Of course that leads to a nasty kind of asymmetry.
We could also finally implement operations with more than 6 arguments.

But for now my personally preferred solution would be to punt on deciding this, and simply not implementing this for more than two indices until somebody explicitly asks for it, with a concrete use case.

Anyway, I am open to all options here :-).

fingolfin commented:
Recent experiments lead me to wonder how `a[i,j]` should behave for out-of-bounds access. I.e. what if `i` or `j` exceeds the number of rows / columns? Right now I am tempted to do this:
* when reading, produce an error (just like `a[i][j]` would)
* when writing, then ...
  * for (p)lists of (p)lists, just emulate `a[i][j]` and add new rows / columns as necessary
  * for "real" matrix objects, also raise an error -- after all, in general this action would result in a non-matrix in general
* `IsBound(a[i,j])` -- I don't consider this very useful, but we might want to support it for "completness" anyway?!?
* `Unbind(a[i,j])` -- again, for (p)list of (p)lists we could support this. But for real matrix objects, I think this should simply raise an error.

ChrisJefferson commented:
I agree with you. For plists, try to make it as interchangeable as possible. For matrices, just don't allow changing the shape.

markuspf commented:
Mhm. Not allowing `Unbind(a[i,j])` on a matrix could make @mohamed-barakat unhappy; I think it'd be perfectly legal for entries of matrices to be "not yet defined", or "not defined".

One could argue that once they have been entered we don't allow Unbinding though.

Other than this I like the approach; we'll have to have an answer to the "why not arbitrarily many indices" question (implementing arbitrary argument methods will be a bit of a pain).

ChrisJefferson commented:
Having unbound values might be a bit weird -- it would stop you having efficient compressed boolean matrices, as usually in GAP you can assume the last element of a list is bound.

markuspf commented:
These are different types of representation. A boolean matrix can (and should) be "total", but a matrix in general might have (and according to @mohamed-barakat must have) the ability to have unbound (or at least lazily computed) entries.

Just outright forbidding unbound entries is just as wrong as assuming that all objects should support unbound entries.

Looks like the matrix objects might gain an extra filter to reflect the fact whether the matrix allows being "partially defined"?

fingolfin commented:
Of course @mohamed-barakat can speak for himself, but I am actually pretty sure / hopeful that `Unbind` is irrelevant for him. And I think in a `MatrixObj` no entry must ever be "unbound" -- however, that does not mean that you have to store the entry anywhere (see: sparse matrices), nor that you have to be know it (see: lazily computed entries); I think it is the latter what Mohamed does. And while he might (or might not) need to (un)bind entries of some underlying data structure, I don't think this is necessary for his outside facing MatrixObj interface.

fingolfin commented:
So if I try to access an "not yet defined" entry, I'd imagine @mohamed-barakat's code would simply start to compute it on the fly.

All in all, I still am firmly convinced that a `MatrixObj` must not have any unbound entries. And note that this doesn't preclude a specific implementation from doing it -- it just means that (a) MatrixObjs implementation need not worry about implementing this, and (b) generic algorithms dealing with MatrixObjs need not worry about catching unbound entries.

ssiccha commented on src/vec8bit.c:
```diff
@@ -5733,7 +5732,7 @@ Obj FuncTRANSPOSED_MAT8BIT( Obj self, Obj mat)
 
                     /* set entry */
-                    ptr = BYTES_VEC8BIT(ELM_GF2MAT(tra, nstart + j)) + imod;
+                    ptr = BYTES_VEC8BIT(ELM_MAT8BIT(tra, nstart + j)) + imod;
```
Then line 4036 (in this commit) can also be changed from using `SET_ELM_GF2MAT` to `SET_ELM_MAT8BIT`.

fingolfin replied:
But in line 4036, a GF2MAT is accessed... Of course internally the two functions do the same, but still.

ssiccha commented on src/lists.h:
```diff
@@ -409,6 +408,18 @@ extern void ASSB_LIST( Obj list, Obj pos, Obj obj );
+*F  ASS2_LIST( <list>, <pos1>, <pos2>, <obj> )
+**
+**  'ASS2_LIST' implements 'list[pos1,pos2]:=obj', which for lists of lists
+**  is defined as 'list[pos1][pos2]', and for other kind of objects is handled
```
Should be `list[pos1][pos2]:=obj`

fingolfin replied:
Thanks, will be fixed

ssiccha commented on src/lists.h:
```diff
@@ -409,6 +408,18 @@ extern void ASSB_LIST( Obj list, Obj pos, Obj obj );
+**
+**  'ASS2_LIST' implements 'list[pos1,pos2]:=obj', which for lists of lists
+**  is defined as 'list[pos1][pos2]', and for other kind of objects is handled
+**  by method dispatch through the GAP attribute 'ELM_LIST' with three
```
operation 'ASS_LIST'

fingolfin commented:
@ssiccha I think I addressed all your comments

fingolfin commented:
This PR is now ready to be merged:  I discussed this PR with @frankluebeck today, and he has no objections. Afterwards, I rebased this one last time, and also removed the benchmark code. We still need to document stuff etc., but that will come separately.
